paddle/framework/program_desc.h

```diff
@@ -45,10 +45,19 @@ class ProgramDesc {
 
   proto::ProgramDesc *Proto();
 
+  // 4 utility functions for inference (TODO: Better comment)
+  const std::vector<std::string> &GetFeedVarNames() const;
+  const std::vector<std::string> &GetFetchVarNames() const;
```

Comment:
I think the two functions are not special for inference. When there are `feed_op` in `ProgramDesc`, we can always call this function to get feed var names. Same to `GetFetchVarNames()`.

I am thinking another implementation, not keeping a member of `feed_var_names_` and traversing all the operators to find the inputs of `feed_op`. Like:

https://github.com/PaddlePaddle/Paddle/blob/9fea1d438563490f2cfe7b00ca283177e4fc75fa/paddle/inference/inference.cc#L43-L52

Which way looks better?

Reply:
Thanks for the nice suggestion. But if go with option 2 (your suggestion), then we will need some placeholder class for keeping these 2 lists right? (For example, `ProgramBuilder` in your initial design). Otherwise, these 2 lists will have to be floating around in the program.
Or maybe i misunderstood?

Reply:
> we will need some placeholder class for keeping these 2 lists right?
 
There is no need to keep the 2 lists. When we want to get the 2 lists, we can always traverse all the operators of the given `ProgramDesc` to get them. It is not time-consuming. We can make sure that the `feed_var_names` and `fetch_var_names` got from the two interfaces are consistent with the user `ProgramDesc`.

Reply:
Sorry, i don't understand this:
> We can make sure that the feed_var_names and fetch_var_names got from the two interfaces are consistent

Reply:
@sidgoyal78  I prefer @Xreki's suggestion because we don't need to add extra members to `ProgramDesc` class.  

In your code, it is possible that the `feed_var_names` data member is inconsistent with the actual info in `ProgramDesc` because you construct the `feed_var_names` using `InsertFetchVarName`.

Following @Xreki's suggestion:
We don't change any code in the `ProgramDesc` class at all, thus we don't need these four functions.
Instead, we provide the general (not a member function of programdesc class) :
```
const std::vector<std::string> &GetFeedVarNames(ProgramDesc*) const;
const std::vector<std::string> &GetFetchVarNames(ProgramDesc*) const;
```
which obtain the vector of names on the fly by traversing the global block. This way we also make sure the obtained names are consistent with program desc.

Reply:
I think in your suggestion InsertFetch and InsertFeed are not taken into account. Maybe we can go away without those inserts in this PR (assuming that the user is not provided with the functionality of inserts).

Reply:
Yes. I don't think InsertFetch and InsertFeed are good utility function because they are error prone. We can provide SetFetch SetFeed function instead to overwrite the feed/fetch info contained in ProgramDesc if user wants to.